sources/api/pluto/src/main.rs

```diff
@@ -260,24 +281,18 @@ async fn get_ipv4_cluster_dns_ip_from_imds_mac(client: &mut ImdsClient) -> Resul
 }
 
 /// Gets the IP address that should be associated with the node.
-async fn get_node_ip(client: &mut ImdsClient) -> Result<String> {
-    // Retrieve the user specified cluster DNS IP if it's specified, otherwise use the pluto-generated
-    // cluster DNS IP
-    let aws_k8s_info = api::get_aws_k8s_info().await.context(error::AwsInfoSnafu)?;
-    let configured_cluster_dns_ip = aws_k8s_info
+async fn generate_node_ip(client: &mut ImdsClient, aws_k8s_info: &mut AwsK8sInfo) -> Result<()> {
+    if aws_k8s_info.node_ip.is_some() {
+        return Ok(());
+    }
+    let cluster_dns_ip = aws_k8s_info
```

Comment:
once the recursion is avoided:
```suggestion
    generate_cluster_dns_ip()?;
    let cluster_dns_ip = aws_k8s_info
```